### PR TITLE
Authority relationship

### LINKED ISSUE #11062: adding relationships to Authority
##### **Overview of the issue**

It is possible to create relationships from user-generated entities to the built-in `User` entity, if a user wants to extend the user without modifying the actual user entity. The same thing is not possible to do with the `Authority` entity, because of:

1. JDL import is not aware of the `Authority` exception. So it either complains about that the `Authority` entity is missing. If we add an `entity Authority {}`, it completely overwrites the actual authority and breaks the code deeply.
2. You can add 'authority-relations' manually using `jhipster entity` without errors. However, the generarator is not aware of the `Authority` exception either. The generated code is also broken. It seems, that the manual generator does not check if the related entity actually exist

##### **Motivation for or Use Case**
Adding a relationship to a built-in entity is one way of extending the entity without changing its generated code. It has been possible to do so for a long time already for the user. But not for its corresponding authority entity.

In a current use-case, I was designing a system, which is capable of handling a bigger amount of permissions which itself can be categorized in groups. In addition to this, there are a few more extensions on the authority itself. With this issue here, it is not possible to realize this scenario without extensively changing the generated code.

##### **Reproduce the error**

1. with JDL

create the following model:

  <details>
  <summary>JDL definitions</summary>
  <pre>
entity AuthorityGroup {
	name String required
    description String required
}

relationship ManyToMany {
	AuthorityGroup{authority(name)} to Authority
}
  </pre>
  </details>

`jhipster import-jdl` will fail to generate an application because `Authority` is missing. Here is the actual error:
  <details>
  <summary>Errorlog of `jhipster import-jdl app.jdl`</summary>
  <pre>
Error: Can't add invalid relationship. Error: In the Many-to-Many relationship from AuthorityGroup to Authority, only bidirectionality is supported. The injected field in the source entity is 'authority(name)' and the injected field in the destination entity is not set.
Error while parsing applications and entities from the JDL Error: Can't add invalid relationship. Error: In the Many-to-Many relationship from AuthorityGroup to Authority, only bidirectionality is supported. The injected field in the source entity is 'authority(name)' and the injected field in the destination entity is not set.
Error: Can't add invalid relationship. Error: In the Many-to-Many relationship from AuthorityGroup to Authority, only bidirectionality is supported. The injected field in the source entity is 'authority(name)' and the injected field in the destination entity is not set.
  </pre>
  </details>

2. manually

If you add the `AuthorityGroup` entity manually by using `jhipster entity`, the code generation will work without any issues, but the generated code will fail in angular frontend code:

  <details>
  <summary>Errorlog when building frontend code</summary>
  <pre>
ERROR in ./src/main/webapp/app/entities/authority-group/authority-group-update.component.ts
Module not found: Error: Can't resolve 'app/entities/authority/authority.service' in '/tmp/jtest/src/main/webapp/app/entities/authority-group'

ERROR in /tmp/jtest/src/main/webapp/app/entities/authority-group/authority-group-update.component.ts
ERROR in /tmp/jtest/src/main/webapp/app/entities/authority-group/authority-group-update.component.ts(11,28):
TS2307: Cannot find module 'app/shared/model/authority.model'.

ERROR in /tmp/jtest/src/main/webapp/app/entities/authority-group/authority-group-update.component.ts
ERROR in /tmp/jtest/src/main/webapp/app/entities/authority-group/authority-group-update.component.ts(12,34):
TS2307: Cannot find module 'app/entities/authority/authority.service'.

ERROR in /tmp/jtest/src/main/webapp/app/shared/model/authority-group.model.ts
ERROR in /tmp/jtest/src/main/webapp/app/shared/model/authority-group.model.ts(1,28):
TS2307: Cannot find module 'app/shared/model/authority.model'.

  </pre>
  </details>

##### **Suggest a Fix**

Adding similar handling as we do with `User` today.

##### **JHipster Version(s)**

6.6.0
##### **JHipster configuration**

default settings

-   [X] Checking this box is mandatory (this is just to show you read everything)

### PR DESCRIPTION
**Pull request by xetys**
enables relationships to 'Authority' in entity-server and ensures using the correct fields in generated files

fix #11062

-   Please make sure the below checklist is followed for Pull Requests.

-   [ ] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
-   [ ] Tests are added where necessary
-   [ ] Documentation is added/updated where necessary
-   [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

**Comment by xetys** [comment]
hey @jhipster/developers I have finished (IMHO) this feature and I was testing it against this JDL:

```

entity Alpha
entity Beta
entity Gamma


relationship OneToOne {
    Alpha{authority(name)} to Authority
}

relationship ManyToOne {
    Beta{authority(name)} to Authority
}

relationship ManyToMany {
    Gamma{authority(name)} to Authority
}
```
and this works fine...

So I wanted to add some meaningful test to cover it, and I'm struggling with the yeoman helper and asking for some hint to make this work:

in `entity.spec.js` I added a new test case with the following code:

``` javascript
            describe('with authority as relationship', () => {
                before(done => {
                    helpers
                        .run(require.resolve('../generators/entity'))
                        .inTmpDir(dir => {
                            fse.copySync(path.join(__dirname, '../test/templates/default-ng2'), dir);
                        })
                        .withArguments(['foo'])
                        .withPrompts({
                            fieldAdd: false,
                            relationshipAdd: true,
                            otherEntityName: 'authority',
                            relationshipName: 'authority',
                            relationshipType: 'many-to-one',
                            ownerSide: true,
                            otherEntityRelationshipName: 'foo',
                            otherEntityField: 'name'
                        })
                        .withPrompts({
                            relationshipAdd: false,
                            dto: 'no',
                            service: 'no',
                            pagination: 'no'
                        })
                        .on('end', done);
                });

                it('creates expected default files', () => {
                    assert.file(expectedFiles.server);
                    assert.file(expectedFiles.clientNg2);
                    assert.file(expectedFiles.gatling);
                    assert.file(expectedFiles.fakeData);
                });

                it('creates correct entity files for authority', () => {
                    // entity related files exist
                    assert.file(`${CLIENT_MAIN_SRC_DIR}app/entities/foo/foo.module.ts`);
                    assert.file(`${CLIENT_MAIN_SRC_DIR}app/entities/foo/foo.component.html`);
                    assert.file('.jhipster/Foo.json');
                    assert.fileContent('.jhipster/Foo.json', 'many-to-one');
                    assert.fileContent('.jhipster/Foo.json', {
                        relationships: [
                            {
                                relationshipType: 'many-to-one',
                                otherEntityName: 'authority',
                                otherEntityRelationshipName: 'foo',
                                relationshipName: 'authority',
                                otherEntityField: 'name'
                            }
                        ]
                    });
                    assert.fileContent(
                        `${CLIENT_MAIN_SRC_DIR}app/entities/foo/foo.component.html`,
                        "<a [routerLink]=\"['/authority', foo.authority?.name, 'view']\" >{{ foo.authority?.name }}</a>"
                    );
                });
            });

```

I try to generate an entity **with** a relationship. From my dumb perspective, this should be the correct usage of Yeomans test helper to answer the prompts. However, this ends up with an entity without a relationship. What am I doing wrong here?

**Comment by xetys** [comment]
push, I still need some help with the test, then this PR could be merged

**Comment by MathieuAA** [comment]
I'll try to see what's wrong this morning

**Comment by xetys** [comment]
thanks... I dunno what's wrong with the usage of the helper...and it's quite complex because the relationship question is a repeating one....if that's not possible, I'll try to define a `.yo-rc.json` and entity JSONs and do it this way

**Comment by MathieuAA** [comment]
@xetys it may be easier indeed

File: test/entity.spec.js [hunk 1]
```diff
@@ -346,6 +346,64 @@ describe('JHipster generator for entity', () => {
                     assert.jsonFileContent('.jhipster/Foo.json', { clientRootFolder: 'test-root' });
                 });
             });
+
+            describe('with authority as relationship', () => {
+                before(done => {
+                    helpers
+                        .run(require.resolve('../generators/entity'))
+                        .inTmpDir(dir => {
+                            fse.copySync(path.join(__dirname, '../test/templates/default-ng2'), dir);
```
Comment (mshima) [inline test/entity.spec.js]: You set authority relationship as oauth2 only.
 You should move to a oauth2 template or enabling authority relationship with session.

File: generators/client/files-angular.js [hunk 2]
```diff
@@ -346,6 +346,11 @@ const files = {
             path: ANGULAR_DIR,
             templates: ['core/user/user.service.ts', 'core/user/user.model.ts']
         },
+        {
+            condition: generator => !generator.skipUserManagement || generator.authenticationType === 'oauth2',
```
Comment (mshima) [inline generators/client/files-angular.js]: See tests case.

File: generators/client/templates/angular/src/main/webapp/app/core/user/user.model.ts.ejs [hunk 3]
```diff
@@ -18,35 +18,35 @@
...
-    activated?: boolean;
-    langKey?: string;
-    authorities?: string[];
-    createdBy?: string;
-    createdDate?: Date;
-    lastModifiedBy?: string;
-    lastModifiedDate?: Date;
-    password?: string;
+id?: any;
+login?: string;
```
Comment (mshima) [inline generators/client/templates/angular/src/main/webapp/app/core/user/user.model.ts.ejs]: The indentation was fine. Why the changes?

File: generators/generator-base-private.js [hunk 4]
```diff
@@ -1223,7 +1235,9 @@ module.exports = class extends Generator {
                 const importType = `I${otherEntityAngularName}`;
                 let importPath;
                 if (otherEntityAngularName === 'User') {
-                    importPath = clientFramework === ANGULAR ? 'app/core/user/user.model' : 'app/shared/model/user.model';
+                    importPath = clientFramework === 'angularX' ? 'app/core/user/user.model' : 'app/shared/model/user.model';
```
Comment (mshima) [inline generators/generator-base-private.js]: You should use the constant here.

File: generators/entity/prompts.js [hunk 5]
```diff
@@ -983,7 +983,7 @@ function askForRelationship(done) {
                         name: 'one-to-one'
                     }
                 ];
-                if (response.otherEntityName.toLowerCase() !== 'user') {
+                if (['user', 'authority'].includes(response.otherEntityName.toLowerCase())) {
```
Comment (MathieuAA) [inline generators/entity/prompts.js]: Are you sure? It previously meant "if the other entity's name isn't user", and now it means "if the other entity's name is either user or authority". Haven't you forgotten a `!`?

File: generators/entity-server/templates/src/main/java/package/domain/Entity.java.ejs [hunk 6]
```diff
@@ -492,6 +493,7 @@ public class <%= asEntity(entityClass) %> implements Serializable {
         const otherEntityNamePlural = relationships[idx].otherEntityNamePlural;
         const otherEntityRelationshipNameCapitalized = relationships[idx].otherEntityRelationshipNameCapitalized;
         const otherEntityRelationshipNameCapitalizedPlural = relationships[idx].otherEntityRelationshipNameCapitalizedPlural;
+        const isOtherEntityFromUserManagement = asEntity(otherEntityNameCapitalized) === 'User' || asEntity(otherEntityNameCapitalized) === 'Authority';
```
Comment (MathieuAA) [inline generators/entity-server/templates/src/main/java/package/domain/Entity.java.ejs]: `isOtherEntityFromUserManagement` -> `userManagementEntity`
I initially thought it was a function (like isBlue) but it's a variable, no need to use a verb here

**Comment by mraible** [comment]
@xetys FYI... this PR has conflicts that need to be resolved.

File: generators/client/files-angular.js (hunk 2 above)
Comment (xetys) [reply generators/client/files-angular.js:350]: seems to be some IntelliJ fuu

File: generators/client/templates/angular/src/main/webapp/app/core/user/user.model.ts.ejs (hunk 3 above)
Comment (xetys) [reply generators/client/templates/angular/src/main/webapp/app/core/user/user.model.ts.ejs:22]: seems to be some IntelliJ fuu

File: generators/entity/prompts.js (hunk 5 above)
Comment (xetys) [reply generators/entity/prompts.js:986]: yeah I did >.<

File: generators/entity-server/templates/src/main/resources/config/liquibase/changelog/added_entity.xml.ejs [hunk 7]
```diff
@@ -128,19 +128,25 @@
                 let nullable_relation = true,
                 relationshipType = relationships[idx].relationshipType,
                 relationshipName = relationships[idx].relationshipName,
-                relationshipColumnType = relationships[idx].otherEntityName === 'user' && authenticationType === 'oauth2' ? 'varchar(100)' : 'bigint';
+                idSuffix = relationships[idx].otherEntityName.toLowerCase() === 'authority' ? 'name' : 'id';
```
[the remark this answers is not in the document]
Comment (xetys) [reply generators/entity-server/templates/src/main/resources/config/liquibase/changelog/added_entity.xml.ejs:131]: This is a quite edgy case, as all 'usual' relationships are going over the 'id' (even if the id is not an int, like user in oauth). Authority is the only exception for this case

File: generators/entity/prompts.js (hunk 5 above)
Comment (MathieuAA) [reply generators/entity/prompts.js:986]: ;)

**Comment by xetys** [comment]
@mraible thx...I didn't forget this one

now I fixed the test as well, and all stuff mentioned in Review (hopefully)

May this PR finally be closed :pray:

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-detail.component.html.ejs [hunk 8]
```diff
@@ -64,6 +64,7 @@
                     const otherEntityStateName = relationships[idx].otherEntityStateName;
                     const otherEntityField = relationships[idx].otherEntityField;
                     const otherEntityFieldCapitalized = relationships[idx].otherEntityFieldCapitalized;
+                    const otherEntityId = otherEntityName === 'authority' ? 'name' : 'id';
```
Comment (mshima) [inline generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-detail.component.html.ejs:67]:
```suggestion
                    const otherEntityId = _.upperFirst(relationships[idx].otherEntityFieldName);
```

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-update.component.ts.ejs [hunk 9]
```diff
@@ -324,22 +346,51 @@ _%>
         this.isSaving = false;
     }
 
-    <%_ if (selectableEntitiesType) { _%>
+    <%_ if (selectableEntitiesType) {
+        const idColumn = selectableEntitiesType === 'IAuthority' ? 'name' : 'id';
```
Comment (mshima) [inline generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-update.component.ts.ejs:350]: Use otherEntityFieldName somehow?

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management.component.html.ejs [hunk 10]
```diff
@@ -116,8 +116,13 @@
                     const relationshipFieldNamePlural = relationships[idx].relationshipFieldNamePlural;
                     const otherEntityName = relationships[idx].otherEntityName;
                     const otherEntityStateName = relationships[idx].otherEntityStateName;
-                    const otherEntityField = relationships[idx].otherEntityField;
+                    let otherEntityField = relationships[idx].otherEntityField;
                     const otherEntityFieldCapitalized = relationships[idx].otherEntityFieldCapitalized;
+                    const idSuffix = otherEntityName === 'authority' ? 'name' : 'id';
```
Comment (mshima) [inline generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management.component.html.ejs:121]:
```suggestion
                    const idSuffix = relationships[idx].otherEntityFieldName;
```

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management.component.html.ejs [hunk 11]
```diff
@@ -116,8 +116,13 @@
...
+                    // fallback for authority, if user has forgotten to define name....authorities do not have an ID
```
Comment (mshima) [inline generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management.component.html.ejs:122]: This fallback isn't necessary, or it should be done at the generator.

File: generators/entity-server/templates/src/main/java/package/domain/Entity.java.ejs [hunk 12]
```diff
@@ -282,6 +282,7 @@ public class <%= asEntity(entityClass) %> implements Serializable {
         const otherEntityNameCapitalized = relationships[idx].otherEntityNameCapitalized;
         const ownerSide = relationships[idx].ownerSide;
         const isUsingMapsId = relationships[idx].useJPADerivedIdentifier;
+        const idSuffix = relationships[idx].otherEntityTableName.toLowerCase() === 'authority' ? 'name' : 'id';
```
Comment (mshima) [inline generators/entity-server/templates/src/main/java/package/domain/Entity.java.ejs:285]:
```suggestion
        const idSuffix = relationships[idx].otherEntityFieldName;
```

File: generators/entity-server/templates/src/main/java/package/service/EntityQueryService.java.ejs [hunk 13]
```diff
@@ -134,10 +134,13 @@ public class <%= serviceClassName %> extends QueryService<<%= asEntity(entityCla
             });
 
             relationships.forEach((relationship) => {
-                const metamodelFieldName = (relationship.relationshipType === 'many-to-many' || relationship.relationshipType === 'one-to-many') ? relationship.relationshipFieldNamePlural : relationship.relationshipFieldName; _%>
-            if (criteria.get<%= relationship.relationshipNameCapitalized %>Id() != null) {
-                specification = specification.and(buildSpecification(criteria.get<%= relationship.relationshipNameCapitalized %>Id(),
-                    root -> root.join(<%= asEntity(entityClass) %>_.<%= metamodelFieldName %>, JoinType.LEFT).get(<%= asEntity(relationship.otherEntityNameCapitalized) %>_.id)));
+                const metamodelFieldName = (relationship.relationshipType === 'many-to-many' || relationship.relationshipType === 'one-to-many') ? relationship.relationshipFieldNamePlural : relationship.relationshipFieldName;
+                const idColumn = relationship.otherEntityNameCapitalized === 'Authority' ? 'name' : 'id';
```
Comment (mshima) [inline generators/entity-server/templates/src/main/java/package/service/EntityQueryService.java.ejs:138]:
```suggestion
                const idColumn = relationships[idx].otherEntityFieldName;
```

File: generators/entity-server/templates/src/main/java/package/service/EntityQueryService.java.ejs [hunk 14]
```diff
@@ -134,10 +134,13 @@ public class <%= serviceClassName %> extends QueryService<<%= asEntity(entityCla
...
+                const idColumnCapitalized = idColumn === 'name' ? 'Name' : 'Id';
```
Comment (mshima) [inline generators/entity-server/templates/src/main/java/package/service/EntityQueryService.java.ejs:139]:
```suggestion
                const idColumnCapitalized = _.upperFirst(idColumn);
```

File: generators/entity-server/templates/src/main/java/package/service/dto/EntityCriteria.java.ejs [hunk 15]
```diff
@@ -58,13 +58,19 @@ relationships.forEach((relationship) => {
 const relationshipType = relationship.relationshipType;
 // user has a String PK when using OAuth, so change relationships accordingly
 let oauthAwareReferenceFilterType = referenceFilterType;
+let idColumn = 'Id';
```
Comment (mshima) [inline generators/entity-server/templates/src/main/java/package/service/dto/EntityCriteria.java.ejs:61]:
```suggestion
let idColumn = _.upperFirst(relationships.otherEntityFieldName);
```

File: generators/entity-server/templates/src/main/resources/config/liquibase/changelog/added_entity_constraints.xml.ejs [hunk 16]
```diff
@@ -28,35 +28,36 @@
         <% for (idx in relationships) {
             const relationshipType = relationships[idx].relationshipType,
             relationshipName = relationships[idx].relationshipName,
+            idSuffix = relationships[idx].otherEntityTableName.toLowerCase() === 'authority' ? 'name' : 'id',
```
Comment (mshima) [inline generators/entity-server/templates/src/main/resources/config/liquibase/changelog/added_entity_constraints.xml.ejs:31]:
```suggestion
            idSuffix = relationships[idx].otherEntityFieldName,
```

File: generators/entity-server/templates/src/main/java/package/service/EntityQueryService.java.ejs [hunk 17]
```diff
@@ -134,10 +134,13 @@ public class <%= serviceClassName %> extends QueryService<<%= asEntity(entityCla
             });
 
             relationships.forEach((relationship) => {
-                const metamodelFieldName = (relationship.relationshipType === 'many-to-many' || relationship.relationshipType === 'one-to-many') ? relationship.relationshipFieldNamePlural : relationship.relationshipFieldName; _%>
-            if (criteria.get<%= relationship.relationshipNameCapitalized %>Id() != null) {
-                specification = specification.and(buildSpecification(criteria.get<%= relationship.relationshipNameCapitalized %>Id(),
-                    root -> root.join(<%= asEntity(entityClass) %>_.<%= metamodelFieldName %>, JoinType.LEFT).get(<%= asEntity(relationship.otherEntityNameCapitalized) %>_.id)));
+                const metamodelFieldName = (relationship.relationshipType === 'many-to-many' || relationship.relationshipType === 'one-to-many') ? relationship.relationshipFieldNamePlural : relationship.relationshipFieldName;
+                const idColumn = relationships[idx].otherEntityFieldName;
```
Comment (mshima) [inline generators/entity-server/templates/src/main/java/package/service/EntityQueryService.java.ejs:138]:
```suggestion
                const idColumn = relationships.otherEntityFieldName;
```

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management.component.html.ejs (hunk 11 above)
Comment (xetys) [reply generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management.component.html.ejs:122]: moved to entity index

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-update.component.ts.ejs (hunk 9 above)
Comment (xetys) [reply generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-update.component.ts.ejs:350]: not sure if I should add more time here, as in this particular template we are working with a lot of variables anywhere. Maybe it's a concern to do it this way and adding another variable doesn't make it better. But I fear to make it in a different way, we should do a bigger work on here. WDYT?

Comment (mshima) [reply generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-update.component.ts.ejs:350]: Yep, looks like we need a bigger refactoring here.

File: generators/entity-server/templates/src/main/resources/config/liquibase/changelog/added_entity_constraints.xml.ejs [hunk 18]
```diff
@@ -28,35 +28,36 @@
         <% for (idx in relationships) {
             const relationshipType = relationships[idx].relationshipType,
             relationshipName = relationships[idx].relationshipName,
+            idSuffix = relationships[idx].otherEntityFieldName,
             ownerSide = relationships[idx].ownerSide,
             otherEntityTableName = relationships[idx].otherEntityTableName;
-            if (relationshipType === 'many-to-one' || (relationshipType === 'one-to-one' && ownerSide)) {
+        if (relationshipType === 'many-to-one' || (relationshipType === 'one-to-one' && ownerSide)) {
```
Comment (mshima) [inline generators/entity-server/templates/src/main/resources/config/liquibase/changelog/added_entity_constraints.xml.ejs:34]:
```suggestion
            if (relationshipType === 'many-to-one' || (relationshipType === 'one-to-one' && ownerSide)) {
```
Ident was right.

File: generators/entity/index.js [hunk 19]
```diff
@@ -994,9 +994,22 @@ class EntityGenerator extends BaseBlueprintGenerator {
...
+                    relationship.otherEntityFieldName = 'id';
                     if (otherEntityName === 'user') {
                         relationship.otherEntityTableName = `${jhiTablePrefix}_user`;
                         hasUserField = true;
+                    } else if (otherEntityName === 'authority') {
+                        relationship.otherEntityTableName = `${jhiTablePrefix}_authority`;
+                        relationship.otherEntityFieldName = 'name';
+                        if (relationship.otherEntityField === 'id') {
+                            this.warning(
+                                chalk.yellow(
```
Comment (mshima) [inline generators/entity/index.js:1006]: Remove chalk here.
Warning already does that.
https://github.com/jhipster/generator-jhipster/blob/0e4b8dc74087603bc781ce2d9983251722af0314/generators/generator-base.js#L1598-L1599

File: generators/entity/index.js [hunk 20]
```diff
@@ -994,9 +994,22 @@ class EntityGenerator extends BaseBlueprintGenerator {
...
                     }
 
+                    relationship.otherEntityFieldName = 'id';
                     if (otherEntityName === 'user') {
                         relationship.otherEntityTableName = `${jhiTablePrefix}_user`;
                         hasUserField = true;
+                    } else if (otherEntityName === 'authority') {
+                        relationship.otherEntityTableName = `${jhiTablePrefix}_authority`;
+                        relationship.otherEntityFieldName = 'name';
+                        if (relationship.otherEntityField === 'id') {
```
Comment (mshima) [inline generators/entity/index.js:1004]: I think otherEntityField is used for bi-directional relationships.
Is this needed?

File: generators/client/templates/angular/src/main/webapp/app/core/user/authority.model.ts.ejs [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+-%>
+
+export interface IAuthority {
+name?: string;
```
Comment (mshima) [inline generators/client/templates/angular/src/main/webapp/app/core/user/authority.model.ts.ejs:21]: Ident.
```suggestion
    name?: string;
```

File: generators/client/templates/angular/src/main/webapp/app/core/user/authority.model.ts.ejs [hunk 22]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+export class Authority implements IAuthority {
+constructor(
+public name?: string,
+) {}
```
Comment (mshima) [inline generators/client/templates/angular/src/main/webapp/app/core/user/authority.model.ts.ejs:27]: Ident
```suggestion
    constructor(
        public name?: string,
    ) {}
```

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management.component.html.ejs [hunk 23]
```diff
@@ -116,8 +116,9 @@
                     const relationshipFieldNamePlural = relationships[idx].relationshipFieldNamePlural;
                     const otherEntityName = relationships[idx].otherEntityName;
                     const otherEntityStateName = relationships[idx].otherEntityStateName;
-                    const otherEntityField = relationships[idx].otherEntityField;
+                    let otherEntityField = relationships[idx].otherEntityField;
```
Comment (mshima) [inline generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management.component.html.ejs:119]:
```suggestion
                    const otherEntityField = relationships[idx].otherEntityField;
```

File: generators/entity-server/templates/src/main/java/package/service/dto/EntityCriteria.java.ejs [hunk 24]
```diff
@@ -58,13 +58,19 @@ relationships.forEach((relationship) => {
...
 // user has a String PK when using OAuth, so change relationships accordingly
 let oauthAwareReferenceFilterType = referenceFilterType;
+let idColumn = _.upperFirst(relationship.otherEntityFieldName);
 if (relationship.otherEntityName === 'user' && authenticationType === 'oauth2') {
     oauthAwareReferenceFilterType = 'StringFilter';
 }
+if (relationship.otherEntityName === 'authority') {
+    oauthAwareReferenceFilterType = 'StringFilter';
+    idColumn = 'Name';
+}
```
Comment (mshima) [inline generators/entity-server/templates/src/main/java/package/service/dto/EntityCriteria.java.ejs:68]:
```suggestion
```

File: generators/entity-server/templates/src/main/java/package/service/dto/EntityCriteria.java.ejs [hunk 25]
```diff
@@ -58,13 +58,19 @@ relationships.forEach((relationship) => {
 const relationshipType = relationship.relationshipType;
 // user has a String PK when using OAuth, so change relationships accordingly
 let oauthAwareReferenceFilterType = referenceFilterType;
+let idColumn = _.upperFirst(relationship.otherEntityFieldName);
 if (relationship.otherEntityName === 'user' && authenticationType === 'oauth2') {
```
Comment (mshima) [inline generators/entity-server/templates/src/main/java/package/service/dto/EntityCriteria.java.ejs:62]:
```suggestion
if (relationship.otherEntityName === 'authority' || (relationship.otherEntityName === 'user' && authenticationType === 'oauth2')) {
```
And probably oauthAwareReferenceFilterType needs a better name.

**Comment by mshima** [comment]
A e2e test is missing too.
Add a TestAuthorityRelationship to https://github.com/jhipster/generator-jhipster/tree/master/test-integration/samples/.jhipster.

And insert here: https://github.com/jhipster/generator-jhipster/blob/0e4b8dc74087603bc781ce2d9983251722af0314/test-integration/scripts/11-generate-entities.sh#L117-L118

File: generators/entity/index.js (hunk 20 above)
Comment (xetys) [reply generators/entity/index.js:1004]: it is for the displayed field in the client. If you don't specify this one in the subgenerator or JDL, say

```
relationship ManyToOne {
 A{authority} to Authority
}
```

we assume 'id' to be the display field. This will lead to compile errors. So that's way this fall back.

**Comment by mshima** [comment]
@xetys have you noticed the failures are related to your PR?

**Comment by xetys** [comment]
Yeah, it's now related to react. As I have no react experience I'll need some time to figure out how to fix this. Shouldn't be too hard but I'll need a bit of time for this

**Comment by xetys** [comment]
still failing because oauth2 is an edge case....I see this is one of the nastiest $100 bug bounties in history :D I'll try to fix that one today >.<

**Comment by mshima** [comment]
> still failing because oauth2 is an edge case....I see this is one of the nastiest $100 bug bounties in history :D I'll try to fix that one today >.<

I don’t think it is required, just throw an exception when at the generator when you detect oauth2.
And blacklist authority from oauth builds, I will think how to do this one.

**Comment by mshima** [comment]
The only way I could find is to duplicate the sqlfull https://github.com/jhipster/generator-jhipster/blob/d4880e95bf6610894493844d843140b59d0e833b/test-integration/scripts/11-generate-entities.sh#L90
@xetys it will be hard to add oauth2 support?

**Comment by mraible** [comment]
@xetys Is this something you're still interested in adding? If so, please fix conflicts.

**Comment by mraible** [comment]
Closing because this PR is stale.

**Comment by Rapster** [comment]
Any news about this? Since https://github.com/jhipster/generator-jhipster/issues/1995 has been closed, I'm thinking to create a custom Authority. If this PR is rejected, then obviously I could do everything manually, but ideally, I would like to design the whole domain with JDL

**Comment by mraible** [comment]
Closing because this PR has become stale. Please re-open if you're able to rebase.
